ui: Fix position of Linked Guideline on graphs

mrtracy opened the pull request:
On our current cluster overview page, we display a "linked guideline" across all
line graphs on the page whenever the user hovers their mouse over any graph;
this allows the user to visually correlate the guideline on the current graph
(provided by NVD3) with the same position on other graphs visible on the page.
However, the previous implementation's linked guideline was slightly out of
position, as well as too short.

This fix changes the method used to display the linked guideline; the new method
uses pure d3 to place a new line directly into the SVG structure rendered by
NVD3; this allows us to use translational "wrapper" elements already created by
NVD3, and places the linked guideline with exact accuracy.  This method is also
more efficient, because it no longer needs to consult the rendered DOM to get
coordinates; instead, attributes can be copied directly from the interactive
guideline element of the active chart.

The new method also maintains no state in the charts; previously, a css class
was used to hide the linked guideline when no graphs are hovering. The new
method completely removes the linked guidelines using d3's exit() function.

Fixes #13001

![screen shot 2017-03-08 at 7 17 57 pm](https://cloud.githubusercontent.com/assets/6969858/23731565/63b8e70a-043c-11e7-8cb4-08ade7030862.png)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14008)
<!-- Reviewable:end -->

Commit:
ui: Fix position of Linked Guideline on graphs

On our current cluster overview page, we display a "linked guideline" across all
line graphs on the page whenever the user hovers their mouse over any graph;
this allows the user to visually correlate the guideline on the current graph
(provided by NVD3) with the same position on other graphs visible on the page.
However, the previous implementation's linked guideline was slightly out of
position, as well as too short.

This fix changes the method used to display the linked guideline; the new method
uses pure d3 to place a new line directly into the SVG structure rendered by
NVD3; this allows us to use translational "wrapper" elements already created by
NVD3, and places the linked guideline with exact accuracy.  This method is also
more efficient, because it no longer needs to consult the rendered DOM to get
coordinates; instead, attributes can be copied directly from the interactive
guideline element of the active chart.

The new method also maintains no state in the charts; previously, a css class
was used to hide the linked guideline when no graphs are hovering. The new
method completely removes the linked guidelines using d3's exit() function.

Fixes #13001

tamird commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/components/graphs.tsx, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFIdWrHXXeDY9pHaO:-KeoFIdWrHXXeDY9pHaP:b-l2i012) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L548)):*
> ```XML
> export function graphMouseMove(graph: SVGElement) {
>   // Select the interactive guideline for the current graph.
>   let currentGuideline = d3.select(graph).select("line.nv-guideline");
> ```

i think a bunch of these lets can be consts (and we should get into the habit of using that feature)

---

*[pkg/ui/app/components/graphs.tsx, line 558 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoG9bSr__aGadjogi-:-KeoG9bSr__aGadjogi0:be7i9qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L558)):*
> ```XML
>     // Linked guideline is inserted into the "nv-wrap" element of the nvd3
>     // graph. This element has the correct translations applied to it, necessary
>     // to display the linked guideline at the correct position.
> ```

expand this comment; if the nv-wrap element is not present, we bail.

also, please audit all the comments in this function; ISTM that this function currently always removes and redraws the linked guideline, despite various comment that indicate something lazier is being done.

---

*[pkg/ui/app/components/graphs.tsx, line 564 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFWPVfrlyPoGsorQE:-KeoFWPWH-NIonxHtZf9:bmphugw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L564)):*
> ```XML
>     }
> 
>     let data = !currentGuideline.empty() ? [currentGuideline] : [];
> ```

can't you hoist this out of this function?

---

*[pkg/ui/app/components/graphs.tsx, line 568 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFjbDD6zKbQAs-Zp5:-KeoFjbDD6zKbQAs-Zp6:bdjkoyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L568)):*
> if the current graph is not displaying an nvd3 guideline

where is this condition? this comment seems out of place

---

*[pkg/ui/app/components/graphs.tsx, line 572 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFsxN_03MHfx32tGg:-KeoFsxN_03MHfx32tGh:b-57bafh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L572)):*
> ```XML
>     container.exit().remove();
> 
>     // If not already present, add the linked guideline to this graph.
> ```

another commented condition which is not present in the code.

---

*[pkg/ui/app/components/graphs.tsx, line 580 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoG3_OLlmjB2HH_Bby:-KeoG3_OLlmjB2HH_Bbz:b-k9ha7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L580)):*
> ```XML
> 
>     // Update linked guideline to match the attributes of the current guideline.
>     let graphLines = container.select(".linked-guideline__line");
> ```

isn't this the return value of the expression just above?

---

*[pkg/ui/app/components/linegraph.tsx, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoEMwRdd772DenzUPU:-KeoEMwRdd772DenzUPV:b-km1zte) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/linegraph.tsx#L28)):*
> ```XML
> export class LineGraph extends React.Component<LineGraphProps, {}> {
>   // The SVG Element in the DOM used to render the graph.
>   graphEl: SVGElement;
> ```

El is short for element? how about graphElement?

---

*[pkg/ui/styl/components/visualizations.styl, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoE07YEF7BDwAvOIx_:-KeoE07YEF7BDwAvOIxa:b-sfej2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/styl/components/visualizations.styl#L80)):*
> ```Stylus
> 
> .linked-guideline__line
>   stroke $glow-blue
> ```

missing newline (you should configure your editor to do this automatically)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/ui/app/components/graphs.tsx, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFIdWrHXXeDY9pHaO:-KeoW-imIrLSnGfwQZAN:bf60pn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L548)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think a bunch of these lets can be consts (and we should get into the habit of using that feature)
</blockquote></details>

I feel like using consts in these cases are more trouble than it's worth. There are few classes that are used in more than one place, and very few that are used widely (in our current code, I would say none). We have no good way to share those with CSS, and we have other techniques that can cut down on duplication better - for example, we use pure-function components in several places for commonly repeated elements.

---

*[pkg/ui/app/components/graphs.tsx, line 558 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoG9bSr__aGadjogi-:-KeoW_NSHTE8OBymsPAd:b-j7zes3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L558)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

expand this comment; if the nv-wrap element is not present, we bail.

also, please audit all the comments in this function; ISTM that this function currently always removes and redraws the linked guideline, despite various comment that indicate something lazier is being done.
</blockquote></details>

Expanded to mention the missing wrapper case.

The "enter/update/exit" pattern is fundamental to D3, so I won't explain that here in much detail; however, the way that we are using it with only a single piece of data (we either have a source guideline or not) might be a bit unintuitive, so I have expanded the comments here significantly.

I've also renamed this function to "updateLinkedGuidelines()", which should make the purpose more clear.

---

*[pkg/ui/app/components/graphs.tsx, line 564 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFWPVfrlyPoGsorQE:-KeoWrkuM9VDdwEzpc9f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L564)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can't you hoist this out of this function?
</blockquote></details>

Done.

---

*[pkg/ui/app/components/graphs.tsx, line 568 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFjbDD6zKbQAs-Zp5:-KeoXDadRhIHZedCLcLt:bj8w3fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L568)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> if the current graph is not displaying an nvd3 guideline

where is this condition? this comment seems out of place
</blockquote></details>

I've overhauled the comments considerably.

---

*[pkg/ui/app/components/graphs.tsx, line 572 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFsxN_03MHfx32tGg:-KeoamFLHjdyvG8-eZrQ:b-azq7hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L572)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

another commented condition which is not present in the code.
</blockquote></details>

See above, overhauled considerably.

---

*[pkg/ui/app/components/graphs.tsx, line 580 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoG3_OLlmjB2HH_Bby:-KeoaHhLXb5bDdQmWgTd:b9a7dt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L580)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

isn't this the return value of the expression just above?
</blockquote></details>

No, the enter() statement is only executed if the guideline is not already present.  "Enter/update/exit" is a fundamental pattern in d3, but I have also updated the comments a bit which should make it more clear even if you don't fully understand that pattern.

---

*[pkg/ui/app/components/linegraph.tsx, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoEMwRdd772DenzUPU:-KeoXRvxY17aOhpD3j9V:b-bn492l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/linegraph.tsx#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

El is short for element? how about graphElement?
</blockquote></details>

use of "el" as shorthand for an element is extremely common in front-end land, going all the way back to backbone.js and probably jQuery.

---

*[pkg/ui/styl/components/visualizations.styl, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoE07YEF7BDwAvOIx_:-Keok2KK5lajfIrqOvWH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/styl/components/visualizations.styl#L80)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

missing newline (you should configure your editor to do this automatically)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008)*
<!-- Sent from Reviewable.io -->

maxlang commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/components/graphs.tsx, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFIdWrHXXeDY9pHaO:-Keowjjdz3hCe3-K0saO:bz58ymo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L548)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I feel like using consts in these cases are more trouble than it's worth. There are few classes that are used in more than one place, and very few that are used widely (in our current code, I would say none). We have no good way to share those with CSS, and we have other techniques that can cut down on duplication better - for example, we use pure-function components in several places for commonly repeated elements.
</blockquote></details>

I think he just meant `const sourceGuideline` instead of `let sourceGuideline` if I understand correctly. 

I feel this should be done all at once rather than arbitrarily enforced on new PRs currently.

---

*[pkg/ui/app/components/graphs.tsx, line 566 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeozSZPY1KyHhi8M9T0:-KeozSZQuXZoFhbXfggs:bb7lbs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d79d31a2b8be466701f662fc4108148a1b77f52e/pkg/ui/app/components/graphs.tsx#L566)):*
> ```XML
>   let otherGraphs = d3.selectAll(".linked-guideline").filter(function () {
>     return this !== hoverGraph;
>   });
> ```

nice

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-:-KeozdO6zkzgWw5O153f:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/components/graphs.tsx, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFIdWrHXXeDY9pHaO:-KeoukkDsnhZYRHFO1gN:bt81a32) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L548)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

I think he just meant `const sourceGuideline` instead of `let sourceGuideline` if I understand correctly. 

I feel this should be done all at once rather than arbitrarily enforced on new PRs currently.
</blockquote></details>

Let's take it easy on the word "enforce"; I'm making a suggestion.

[written before max's comment above]: Sorry, I didn't mean to imply that you should make things that are currently being mutated immutable; I only meant: "if any of these are already immutable, declare them with `const`".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/components/graphs.tsx, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFIdWrHXXeDY9pHaO:-Kep-mlpadPZGE0Oj8nO:bmt5em) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L548)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Let's take it easy on the word "enforce"; I'm making a suggestion.

[written before max's comment above]: Sorry, I didn't mean to imply that you should make things that are currently being mutated immutable; I only meant: "if any of these are already immutable, declare them with `const`".
</blockquote></details>

Ah yeah sorry I see how that might've come off. I actually think it should be enforced so that's why I used the word, but then I reflected that if we just did that going forward it would feel arbitrary. I definitely didn't read your comment as enforcement, just a suggestion which ultimately I agree with beyond the scope of this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/ui/app/components/graphs.tsx, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008#-KeoFIdWrHXXeDY9pHaO:-KepClUmquQw6HcGyibQ:b6s4a8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3983f9a3ef06fd030bfe3c97f4e4e67c2783cdf4/pkg/ui/app/components/graphs.tsx#L548)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

Ah yeah sorry I see how that might've come off. I actually think it should be enforced so that's why I used the word, but then I reflected that if we just did that going forward it would feel arbitrary. I definitely didn't read your comment as enforcement, just a suggestion which ultimately I agree with beyond the scope of this PR.
</blockquote></details>

Ah, I misunderstood, my brain elided the word "lets"; didn't realize this was a `let` vs `const` situation.  Honestly hadn't even considered it, seems like an obvious choice to make and there's no reason to do it all at once.  Changed all of these to `const`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14008)*
<!-- Sent from Reviewable.io -->
